include/pisa/scorer/scorer.hpp

```diff
@@ -11,26 +11,37 @@
 #include "quantized.hpp"
 #include "spdlog/spdlog.h"
 
+struct ScorerParams {
+    ScorerParams(std::string s_name = "default") : name(s_name) {}
```

Comment:
A few comments:
1. There shouldn't be a default, `"default"` doesn't mean anything does it? I imagine you did that so you can initialize it in CLI with something. I suggest just initialize it to `ScoreParams("")` **there** and leave this to require a name.
2. The constructor should be `explicit`.
3. You should move the contents of `s_name`: `name(std::move(s_name))`.
4. Maybe let's expand the name of the argument to `scorer_name`.

Reply:
Okay, thanks. Why do we need the `std::move` in this case? Is there some 'best practice' here?

Reply:
The short answer is that you avoid unnecessary copy.

Your constructor gets `s_name` as value, so it will be destructed after the constructor goes out of scope. If you don't move, then an additional copy will be made for `name` constructor. `std::move` effectively moves out any allocated memory to an rvalue reference, while leaving `s_name` in undetermined but valid "empty" state.

It's important to note that small strings will be stored on stack because of small string optimization, so chances are it will be byte-copied anyway given our short names, but if that happens, `std::move` won't have any disadvantages over a copy.

Another point is that alternatively, we could simply pass by reference to the constructor:

```cpp
ScorerParams(std::string const& s_name) : name(s_name) {}
```

This will avoid making an additional copy as well. However, in this case, we'll always be making one copy, while pass-by-value can avoid making any if the argument is an rvalue itself (like a temporary):

```cpp
ScorerParams params("bm25");
// or
ScorerParams params(std::move(another_variable));
```

Again, in our case this whole thing is not very important simply because any overhead is negligible here. But as you suggested, this is kind of a best practice pattern when holding member values of non-trivially-copyable types.

Also, I believe clang-tidy will complain :)